### ISSUE TITLE
Invocation defined recipe

### ISSUE THREAD

**Issue by commonquail** [openrewrite/rewrite#4258]
We can already run OpenRewrite via the Maven plugin without modifying files up front:

```sh
mvn -U org.openrewrite.maven:rewrite-maven-plugin:run \
  -Drewrite.activeRecipes=org.openrewrite.java.RemoveUnusedImports
```

but sadly [only for recipes that require no configuration][:(]. Recipes that do require configuration have no (canonical) inline invocation form, instead necessitating the creation of a verbose and complex file merely to communicate what is often trivial information to the recipe, and then you even _still_ need the `rewrite.activeRecipes` user property from the example. This severely limits the practical utility of OpenRewrite. Instead of being a hypothetical arbitrary rename-refactor the tool is only really interesting for completely static transformations, or for dynamic transformations that are particularly complex or that have to run in batch mode.

[:(]: https://docs.openrewrite.org/running-recipes/running-rewrite-on-a-maven-project-without-modifying-the-build#running-a-recipe-with-configuration-parameters

## What problem are you trying to solve?

Consider a trivial recipe with a structure comparable to

```yaml
---
type: specs.openrewrite.org/v1beta/recipe
name: com.yourorg.ChangePackageExample
displayName: Rename package name example
recipeList:
  - org.openrewrite.java.ChangePackage:
      oldPackageName: com.yourorg.foo
      newPackageName: com.yourorg.bar
```

Suppose this needs to be run by _many different individuals_ for different values of `oldPackageName` or `newPackageName`. For \<reasons> we can distribute instructions for defining and executing a recipe but distributing a recipe itself is impractical; say, for example, that we cannot know literally every variation of package names, or that the old package name can exist in multiple artifacts and should be turned into numerous divergent package names.

The YAML configuration file works well with distribution when distribution is already a solved problem, but distribution is a difficult problem to solve and cumbersome even when already solved. It requires a distribution channel, which you may not already have, and if you do have one, using it may necessitate a disproportionate amount of bureaucracy namely if you're subject to auditing regulations. Even if distribution is feasible, some recipes are impossible to distribute in practice, like the examples from the previous paragraph.

But I don't see why we couldn't "just" pass the package names as user properties like we can with `rewrite.activeRecipes`.

## Describe the solution you'd like

Imagine I could run something like

```sh
mvn -U org.openrewrite.maven:rewrite-maven-plugin:run \
  -Drewrite.activeRecipes=org.openrewrite.java.ChangePackage \
  -Drewrite.ChangePackage.oldPackageName=com.yourorg.foo \
  -Drewrite.ChangePackage.newPackageName=com.yourorg.bar
```

Generalizing, `-Drewrite.RecipeSimpleName.optionName=value`.

This is a command I can email blast to lots of individuals, each of whom can easily adjust the invocation as necessary with no further effort.

The pattern I propose means you can run each distinct recipe only once per invocation but is acceptable to me. Repeating the command for variations in configuration option values is trivial, and if you need to do that a lot you're probably in a situation where the overhead of the configuration file approach is amortized anyway. An _n_-ary invocation form is theoretically possible, it just doesn't seem worth the implementation complexity or the impact to invocation complexity.

I don't care about the exact format of the user property. It only needs to be reasonably resistant to collision and should feature the option name verbatim for clarity.

## Have you considered any alternatives or workarounds?

On multiple occasions I've discounted OpenRewrite in favour of low level tools like `sed` and `mv` and Python on account of this limitation. I've thought about making tools to generate specialized OpenRewrite configuration files but that has not been worth the effort so far.

**Comment by timtebeek** [comment]
Hi @commonquail ; thanks for explaining clearly what you're after in your use of the tool. Did you already happen to come across [this entry](https://docs.openrewrite.org/reference/faq#is-it-possible-to-pass-arguments-to-a-recipe-from-the-command-line-i-want-to-programmatically-config) into our FAQ?

> **Is it possible to pass arguments to a recipe from the command line? I want to programmatically configure complex recipes.**
> Not right now. This is a particularly difficult problem to address for a couple of reasons:
> - Some recipes can be composed of other recipes which could then include other recipes composed of other recipes and so on.
> - Some recipes can be used multiple times in one recipe with different parameters such as in [this example](https://github.com/openrewrite/rewrite-migrate-java/blob/v2.0.6/src/main/resources/META-INF/rewrite/jakarta-ee-9.yml#L140-L160).
> 
> There is an [open issue](https://github.com/openrewrite/rewrite-maven-plugin/issues/345) for this request that you can +1 or provide feedback on.

You might be interested to have a look at the Moderne CLI: There we have a `mod run` command that's able to take in `--recipe-option` for quick runs against serialized LSTs: https://docs.moderne.io/user-documentation/moderne-cli/cli-reference#mod-run
The CLI also integrates with our [Moderne IntelliJ plugin](https://www.moderne.io/blog/introducing-the-moderne-ide-plugin-for-jetbrains-intellij-idea), such that you can search and run across repositories more easily. There's a link to get started at the end of the blog.

It sounds like the CLI might align well with the experience you're after, while supporting more than just Java Maven and Gradle projects. For the OSS Maven and Gradle plugins such an approach might not work as well, given the time it takes to build a model versus running just a single recipe. With the CLI there's no such delays for multiple quick one-off recipe runs.

Of course there's always the option of wrapping your recipes up into an internal recipe library, based on [our rewrite-recipe-starter](https://github.com/moderneinc/rewrite-recipe-starter). Most companies will have the infrastructure to download a jar from Artifactory or Nexus, using `-Drewrite.recipeArtifactCoordinates=g:a:v`. That won't solve all the use cases you've listed above, but could still help.

**Comment by commonquail** [comment]
> Did you already happen to come across [this entry](https://docs.openrewrite.org/reference/faq#is-it-possible-to-pass-arguments-to-a-recipe-from-the-command-line-i-want-to-programmatically-config) into our FAQ?

Oops, no. I figured something like that would have been referenced in the documentation I linked.

I appreciate why composite recipes are Difficult™. It's why I would have personally just ruled them out of scope.

> It sounds like the CLI might align well with the experience you're after

Not really, no. The appeal of Maven (Gradle) is that its distribution can be presumed. The whole point is that I can't control remote file systems -- delegating to another tool isn't any easier than copying a YAML file, and in fact can be a great deal more difficult for executables.

> Of course there's always the option of wrapping your recipes up into an internal recipe library

Yes, this is the difficulty I alluded to. It's a mechanism, sure, but it is incapable of addressing the case I presented.

**Comment by timtebeek** [comment]
> > Did you already happen to come across [this entry](https://docs.openrewrite.org/reference/faq#is-it-possible-to-pass-arguments-to-a-recipe-from-the-command-line-i-want-to-programmatically-config) into our FAQ?
> 
> Oops, no. I figured something like that would have been referenced in the documentation I linked.

Thanks for the hint! I've [added a link to the FAQ](https://docs.openrewrite.org/running-recipes/running-rewrite-on-a-maven-project-without-modifying-the-build#running-a-recipe-with-configuration-parameters) to the page you linked above.

> I appreciate why composite recipes are Difficult™. It's why I would have personally just ruled them out of scope.

Good that we agree on the the difficulties with composite recipes. The challenge then is to provide those additional options only to standalone recipe runs, and educate users about the difference and limitations. Add to that the limitations with providing good user feedback in a Maven plugin, and the fact that there's no immediate parallel to command-line only runs for our friends using Gradle. Taken altogether that made us choose for now not to add these options to the Maven plugin, but focus on a single user experience we can more easily control with the CLI.

> > It sounds like the CLI might align well with the experience you're after
> 
> Not really, no. The appeal of Maven (Gradle) is that its distribution can be presumed. The whole point is that I can't control remote file systems -- delegating to another tool isn't any easier than copying a YAML file, and in fact can be a great deal more difficult for executables.

Understandable that folks are more likely to have Maven installed as opposed to a new tool.

Have you thought of going the opposite direction? Not send out the commands for them to run, but the results of those commands in a PR? That's an approach that we see working well with users of Moderne, where a centralized platform team can help teams get up to speed with recipe runs, sending out the resulting PRs for review in the teams. Just as a way of looking how you can tackle the problems you're facing with existing tools and workflows, as opposed to trying to fit a particular usage pattern.

> > Of course there's always the option of wrapping your recipes up into an internal recipe library
> 
> Yes, this is the difficulty I alluded to. It's a mechanism, sure, but it is incapable of addressing the case I presented.

That might indeed be the case; it's always difficult to visualize a simplified and redacted use case, and I can understand that you can't share everything you're hoping to use these tools for. I can only then show you the pattenrs that we have seen work well, without knowing if those would fit exactly what you're after.

I hope you're still getting something from this discussion though, and we remain open to suggestions on how to improve.
